hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/types/HazelcastObjectType.java

```diff
@@ -36,16 +37,16 @@ public class HazelcastObjectType extends RelDataTypeImpl {
     private final String name;
     private final boolean nullable;
     /** Modifiable list of fields to support cyclic types. */
-    private List<Field> fields = new ArrayList<>();
+    private List<Field> fields = new NonTraversableList<>("Type fields must be finalized");
```

Comment:
nit: I think it would be simpler to have `boolean finalized` and `assert finalized : "Type not yet ready"` (including `getFieldList`) or `assert !finalized : "Cannot add fields to finalized type"` in those few places that use `fields`. And you would have 1 less object to allocate.

Reply:
>1. return Collections.unmodifiableList(fields) in getFieldList(), or
>2.  implement the finalization logic in NonTraversableList.

3. fail if the object is not yet finalized

`digest` should be calculated when the object is constructed as per contract of base class. We extend this process for UDTs and I would expect that `getFieldList()` invocation is invalid before the object is fully constructed. So plain failing should be sufficient.

Reply:
but this is nit, so you can leave it as is